### PR TITLE
Draft: API overhaul, mango 2.0.0

### PR DESCRIPTION
**Pull request by rcschrg**
This PR overhauls significant parts of the mango API.
- send_message new signature, auto sender information
- send_acl_message deleted
- new package message MangoMessage, which contains all meta data for distribution
- new express API
- container factory methods overhaul

Further:
- Doctests will be introduced
- Examples deleted to improve package maintainability (most examples exist as doc snippet or unit test)
- We update requirements to 3.10

@maurerle @LiebeSHolly @MalinRadtke214 @emiliefro 
- I would love your feedback on the new/modified APIs, for more examples, take a look at the tests in express/test_api.py, in docs/source/migration.rst you can find a summary on the API breaks + some thoughts and reasonings for some of them.

The new way to use mango would be:
```python
class PingPongRole(Role):
    counter: int = 0

    def handle_message(self, content: Any, meta: dict):
        if self.counter >= 5:
            return
        
        self.counter += 1

        if content == "Ping":
            self.context.schedule_instant_message("Pong", sender_addr(meta))
        elif content == "Pong":
            self.context.schedule_instant_message("Ping", sender_addr(meta))

@pytest.mark.asyncio
async def test_activate_pingpong():
    container = create_tcp_container("127.0.0.1:5555")
    ping_pong_agent = agent_composed_of(PingPongRole(), register_in=container)
    ping_pong_agent_two = agent_composed_of(PingPongRole(), register_in=container)

    async with activate(container) as c:
        await c.send_message("Ping", ping_pong_agent.addr, sender_id=ping_pong_agent_two.aid)
        while ping_pong_agent.roles[0].counter < 5:
            await asyncio.sleep(0.01)

    assert ping_pong_agent.roles[0].counter == 5
```

### CODE REVIEW

**Comment by maurerle** [comment]
I like these changes.
From a quick review of the changes without a deeper tryout:

1. one does not need to have a running container when creating the simulation anymore - thats a thing I worked around quite badly in the past, so I am happy to see this.
2. I do like the partial FIPA ACL compliance, so I am happy it was not dropped, but I like that the default is now much simpler
3. I do like that Agents can be created without a container now as well. This means that the agent does not have an aid until it is added to a container (where we can set `suggested_aid`). 
4. I like that Agents have a `context` now as well so that Agent and Role does behave more similar
5. The API has breaking changes so downstream projects might get to know about the release through errors - so it would be nice to keep a compatibility function `send_acl_message` which is marked as deprecated and will be removed in 2.1? Though there are quite many changes, which make this very hard. It's probably fine to be breaking here.
6. Having all examples removed is something I do not like that much, but this is probably temporary until they are readded in a working state?

Great work! :)

**Comment by rcschrg** [comment]
Thanks for your feedback :) I am glad most changes are perceived as positive and valuable :) Regarding the two points:
* Convenience method for send_acl_message: I agree that it is quite a huge break, but I feel it makes sense to just remove it, just because you need to change all send_message calls nonetheless, because the address has to be created differently now and the sender_id and addr as arguments can (and should) be dropped.
* Regarding the examples: I honestly do not see the additional value. Most examples are duplicate to some test or the docs. But as we can't test examples automatically, they are, im my opinion, a risk for the maintainability, and currently I would not consider readding them, in the form they were. However, huge parts of the documentation changes are still to be done, and I am committed to add some of the removed examples if they have an additional value to the docs. There, I will rewrite them to doctests and maybe shorten them.

File: mango/container/core.py [hunk 1]
```diff
@@ -621,9 +124,24 @@ def register_agent(self, agent, suggested_aid: str = None):
             self._no_agents_running = asyncio.Future()
         aid = self._reserve_aid(suggested_aid)
         self._agents[aid] = agent
+        agent._do_register(self, aid)
         logger.debug("Successfully registered agent;%s", aid)
         return aid
 
+    def include(self, agent: A, suggested_aid: str = None) -> A:
```
Comment (jsagerOffis) [inline mango/container/core.py:131]: Is there a particular reason this is called `include` now? It's still `register` in Mango.jl so mildly confusing.

File: docs/source/migration.rst [hunk 2]
```diff
@@ -3,7 +3,25 @@ Migration
...
+* The container factory method create_container has been removed and instead there is one factory method per container type now: create_tcp_container, create_mqtt_container, create_external_container
+* The container factory methods are not async anymore! 
+* The startup for the TCP-server/MQTT-client has been moved to `container.start()`.
+  * Use the async context manager provided by `activate(containers)` as described in the documentation
+  * Background: This has the advantage that calling shutdown is not necessary anymore, in that way it is not possible to forget and ressource leaks are avoided at any time 
+* Consider to use the async context manager provided by run_with_X, these managers will provide suitable containers internally, that way the user does not need to handle container at all
+* Agents are created without container now
+  * This implies that Agents need to be registered explicitly using `register`
```
Comment (jsagerOffis) [inline docs/source/migration.rst:19]: Going by the new examples this should be `include` or `register` here? Clearly the `include` API can do this without me ever typing `register`.

File: docs/source/tutorial.rst [hunk 3]
```diff
@@ -167,11 +167,8 @@ As an additional feature, we will make it possible to manually set the agent of
 
 Next, we set up its ``handle_message`` function. The controller needs to distinguish between two message types:
 The replies to feed_in requests and later the acknowledgements that a new maximum feed_in was set by a pv agent.
-We use the ``performative`` field of the ACL message to do this. We set the ``performative`` field to ``inform``
-for feed_in replies and to ``accept_proposal`` for feed_in change acknowledgements. All messages between containers
-are expected to be ACL Messages (or implement the split_content_and_meta function). Since we do not want to create
-the full ACL object ourselves every time, within this example we always use the convenience method
-``container.send_acl_message``, which also supports setting the acl metadata.
+We use the assign the key ``performative``of the metadata of the message to do this. We set the ``performative`` field to ``inform``
```
Comment (jsagerOffis) [inline docs/source/tutorial.rst:170]: words

File: mango/agent/role.py [hunk 4]
```diff
@@ -522,16 +428,78 @@ def remove_role(self, role: Role):
...
+    To interact with the environment you have to use the context, accessible via :func:Role.context.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the roles internals.
+        !!Care!! the role context is unknown at this point!
+        """
+        self._context = None
+
+    def bind(self, context: RoleContext) -> None:
```
Comment (jsagerOffis) [inline mango/agent/role.py:466]: If this is internal use only, should be `__bind`?

File: mango/container/core.py [hunk 5]
```diff
@@ -541,7 +44,7 @@ def __init__(
         self.loop: asyncio.AbstractEventLoop = loop
 
         # dict of agents. aid: agent instance
-        self._agents: Dict = {}
+        self._agents: dict = {}
```
Comment (jsagerOffis) [inline mango/container/core.py:47]: should type hint be more specific than just `dict` here?

File: mango/container/mp.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
+import asyncio
```
Comment (jsagerOffis) [inline mango/container/mp.py:1]: I just had to scroll quite a bit to realize what `mp` means in this file name... makese sense but should we maybe still call it 'multi_process'?

File: mango/express/api.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    # convert decoded addr list to tuple for hashability
+    return AgentAddress(
+        tuple(meta["sender_addr"])
+        if isinstance(meta["sender_addr"], list)
+        else meta["sender_addr"],
+        meta["sender_id"],
+    )
+
+
+def addr(protocol_part: Any, aid: str) -> AgentAddress:
```
Comment (jsagerOffis) [inline mango/express/api.py:250]: If anything changes with the naming on the previous comment on `AgentAddress` and `addr.addr` change would have to be propagated to this function too.

File: mango/container/external_coupling.py [hunk 8]
```diff
@@ -125,14 +134,14 @@ async def _send_external_message(self, addr, message) -> bool:
         self.message_buffer.append(
             ExternalAgentMessage(
                 time=time.time() - self.current_start_time_of_step + self.clock.time,
-                receiver=addr,
+                receiver=addr.addr,
```
[the remark this answers is not in the document]
Comment (rcschrg) [reply mango/container/external_coupling.py:137]: I also like AgentAddress. I like protocol_addr, but I think the short version is a little bit confusing (protobuf, prototype), and it shouldn't be used that often. Most of the time you would just use the AgentAddress object directly (for sending messages) or the container.addr field.

File: mango/container/core.py (hunk 1 above)
Comment (rcschrg) [reply mango/container/core.py:131]: I will change register_agent to register, and delete include. The idea was to keep register_agent unbroken, but in fact nobody used manual registering anyway as it was included in the Agent constructor.

**Pull request merged**